vertx-sql-client/src/main/java/io/vertx/sqlclient/SqlClient.java

```diff
@@ -43,129 +43,12 @@
-   */
-  @Fluent
-  SqlClient preparedQuery(String sql, Tuple arguments, Handler<AsyncResult<RowSet<Row>>> handler);
-
-  /**
-   * Like {@link #preparedQuery(String, Tuple, Handler)} but returns a {@code Future} of the asynchronous result
-   */
-  Future<RowSet<Row>> preparedQuery(String sql, Tuple arguments);
-
-  /**
-   * Execute the given {@code sql} string using a prepared statement with parameters set in the {@code Tuple}, the asynchronous result is represented as a collection of elements transformed by the provided {@link java.util.stream.Collector}.
-   *
-   * @param sql the prepared query SQL
-   * @param arguments the list of arguments
-   * @param collector the collector
-   * @param handler the handler notified with the execution result
-   * @return a reference to this, so the API can be used fluently
-   */
-  @GenIgnore
-  <R> SqlClient preparedQuery(String sql, Tuple arguments, Collector<Row, ?, R> collector, Handler<AsyncResult<SqlResult<R>>> handler);
-
-  /**
-   * Like {@link #preparedQuery(String, Tuple, Collector, Handler)} but returns a {@code Future} of the asynchronous result
-   */
-  @GenIgnore
-  <R> Future<SqlResult<R>> preparedQuery(String sql, Tuple arguments, Collector<Row, ?, R> collector);
-
-  /**
-   * Execute the given {@code sql} string using a prepared statement with a batch of parameters set in the {@code List}, the asynchronous result is represented as a {@link RowSet}.
-   *
-   * @param sql the prepared query SQL
-   * @param batch the batch of tuples
-   * @param handler the handler notified with the execution result
-   * @return a reference to this, so the API can be used fluently
-   */
-  @Fluent
-  SqlClient preparedBatch(String sql, List<Tuple> batch, Handler<AsyncResult<RowSet<Row>>> handler);
-
-  /**
-   * Like {@link #preparedBatch(String, List, Handler)} but returns a {@code Future} of the asynchronous result
-   */
-  Future<RowSet<Row>> preparedBatch(String sql, List<Tuple> batch);
-
-  /**
-   * Execute the given {@code sql} string using a prepared statement with a batch of parameters set in the {@code List}, the asynchronous result is represented as a collection of elements transformed by the provided {@link java.util.stream.Collector}.
-   *
-   * @param sql the prepared query SQL
-   * @param batch the batch of tuples
-   * @param collector the collector
-   * @param handler the handler notified with the execution result
-   * @return a reference to this, so the API can be used fluently
-   */
-  @GenIgnore
-  <R> SqlClient preparedBatch(String sql, List<Tuple> batch, Collector<Row, ?, R> collector, Handler<AsyncResult<SqlResult<R>>> handler);
-
-  /**
-   * Like {@link #preparedBatch(String, List, Collector, Handler)} but returns a {@code Future} of the asynchronous result
-   */
-  @GenIgnore
-  <R> Future<SqlResult<R>> preparedBatch(String sql, List<Tuple> batch, Collector<Row, ?, R> collector);
+  PreparedQuery<RowSet<Row>> preparedQuery(String sql);
```

Comment:
This `PreparedQuery` isn't actually a **prepared** statement, and the `PREPARE` phase is executed lazily?

Reply:
Yes that's correct. I couldn't think of any use case where a user may need to execute some code between the PREPARE and first execution of a prepared statement

Reply:
you might want to prepare a query before execution to ensure the query is valid

Reply:
e.g you prepare the query in the start phase of the application and then use the prepared query at runtime

Reply:
do you mean to ensure the query is valid or to ensure the connection is valid?
Checking if the query is valid should be something that is done at application development time or at least in a testing env. 
If it's to check if the connection is valid, some drivers have a `SqlConnection#ping` method (which we should make standard IMO)

Reply:
also, if we want to keep the `prepare` method, what would it return? And how would we clearly distinguish that return type from the return type of `preparedQuery`?

Currently the proposed architecture is a 3-tier model that closely resembles JDBC (which is good because it will help people understand it quickly).
```
1) SqlConnection/Pool/Transaction
   2) Query<RESULT_TYPE> / PreparedQuery<RESULT_TYPE>
      3) RESULT_TYPE (one of RowSet, RowStream, Cursor, or COLLECTOR_TYPE)
```
Which conceptually maps nicely to JDBC:
```
1) Connection
  2) Statement / PreparedStatement / CallableStatement
    3) ResultSet
```

If we allow an intermediate `PreparedQuery` object for a PREPARE phase result, it inserts a new (infrequently used) layer between (1) and (2).

Reply:
how about keeping the `prepare` APIs unchanged and replace the `SqlClient` APIs with:
```
public interface SqlClient {
    Statement<RowSet<Row>> createStatement(String sql);

    void close();
}
```

```
public interface Statement<R> {

    <K> Statement<R> addOption(PropertyKind<K> propertyKind);

    <T> Statement<SqlResult<T>> asCollector(Collector<?, ?, T> collector);

    Future<R> query();

    Statement<R> query(Handler<AsyncResult<R>> handler);

    Future<R> execute();

    Statement<R> execute(Handler<AsyncResult<R>> handler);

    Future<R> execute(Tuple params);

    Statement<R> execute(Tuple params, Handler<AsyncResult<R>> handler);

    Future<R> executeBatch(List<Tuple> batch);

    Statement<R> executeBatch(List<Tuple> batch, Handler<AsyncResult<R>> handler);

}
```

In this way the `Statement` interface is only a container for SQL string and could be used for `simple query`, one-shot `prepared query` and `prepared batch`. And the `SqlConnection#prepare` and `PreparedQuery` interface works as before

Reply:
I think the `Query` name is fine, no need to use the `Statement` name instead.

Reply:
@BillyYccc What about `Cursor` and `RowStream` result types that only apply to prepared queries? I think it's important to have the return-type-setting `asXXX()` methods on the interface for consistency. This is the main reason why I had the extended `PreparedQuery` interface in the original proposal. If we can apply Cursor and RowStream results to non-prepared statements, then it wouldn't be an issue.

Also, I think it's important to have consistent naming convention on the query execution methods (i.e. `execute()` and `executeBatch()`), it's not really clear that `query()` is so similar to `execute()`, and having a separate `Query` and `PreparedQuery` interface would alleviate that.